arbor/partition_load_balance.cpp

```diff
@@ -88,20 +88,23 @@ domain_decomposition partition_load_balance(
     // The global gj_connection table after gathering is indexed by gid.
     auto global_gj_connection_table = ctx->distributed->gather_gj_connections(local_gj_connection_table);
 
-    // Modify the global gj_connection table to make all gj_connections bidirectional.
+    // Make all gj_connections bidirectional.
+    std::vector<std::unordered_set<cell_gid_type>> missing_peers(global_gj_connection_table.size());
     for (auto gid: make_span(global_gj_connection_table.size())) {
         const auto& local_conns = global_gj_connection_table[gid];
         for (auto peer: local_conns) {
             auto& peer_conns = global_gj_connection_table[peer];
-            // If gid is not in the peer connection table insert it such that
-            // the vector remains sorted.
-            auto it = std::lower_bound(peer_conns.begin(), peer_conns.end(), gid);
-            if (it == peer_conns.end() || *it != gid) {
-                peer_conns.insert(it, gid);
+            // If gid is not in the peer connection table insert it into the
+            // missing_peers set
+            if (!std::binary_search(peer_conns.begin(), peer_conns.end(), gid)) {
+                missing_peers[peer].insert(gid);
             }
         }
     }
-
+    // Append the missing peers into the global_gj_connections table
+    for (unsigned i = 0; i < global_gj_connection_table.size(); ++i) {
+        std::move(missing_peers[i].begin(), missing_peers[i].end(), std::back_inserter(global_gj_connection_table[i]));
```

Comment:
are the entries in `global_gj_connection_table[i]` supposed to be sorted?
The loop above assumes that they are, which enables `std::binary_search/std::lower_bound`, however `back_inserter` to add missing values is not guaranteed to maintain sorted values.